### ISSUE TITLE
Known security issues

### ISSUE THREAD

**Issue by mkow** [oscarlab/graphene#2108]
This issue lists known security problems with Graphene. Everything we're aware of but didn't have time to fix yet should be listed here. We'll also need to resolve all of them before we can say we're ready for production.

I'll try to keep this list up-to-date.

Critical:
- [ ] Security audit of the old code - we do security reviews of newly created code, but there's a lot of old code which was never properly audited.
- [ ] Audit of defaults and their security implications - We must ensure that it's not possible to use Graphene insecurely by an accident because we had some insecure defaults (or misleadingly named options).
- [ ] Better sanitization of CPU topology information in sysfs (#2105) - We forward a lot of information from untrusted host to the user application with only partial sanitization. User app may trust this data, as it thinks they are coming from the trusted OS kernel and so have exploitable bugs in the parsing logic.
- [ ] Finish rework of the logging system. Right now we leak some secrets in error messages even in release builds, we need to carefully censor them.
- [ ] We need better sanitization of OCALL arguments. See #1236.
- [ ] Cryptography review of protected files implementation. We don't have professional cryptographers on the team, so the implementation still needs to be verified.
- [ ] Cryptography review of attestation-related code (SGX-only).
- [ ] Fix all known memory corruption bugs in Graphene (IPC subsystem, I'm looking at you!).

Probably not-so-critical: (or rather: we don't know any app which would be vulnerable because of these)
- CPUID sanitization (#966) - currently we sanitize everything we know may lead to corruptions in users' apps (like xsave area sizes), but maybe there are more dangerous cpuid leafs? We also allow the values to change in time in some cases, which is usually impossible on real hardware.
- There are tons of subtle attack angles related to differences between the upstream Linux API and what we actually implement or can support on a specific backend. E.g. on SGX1 we can't change memory permissions, so everything is RWX, or that we support flag X but not flag Y in some syscalls (while in Linux support for X implies support for Y). It's quite hard for me to imagine a real-world application which would rely on such things for its correctness (and I mean correctness, not exploitability if another bug is present), but it's still worth noting that such a possibility exists.
- Time readings may be easily spoofed, but unless we won't allow it going backwards then it's hard to imagine an application exploitable by changing system time. There may be some issues with TLS certificate validation logic in applications (e.g. the attacker shifts back time and shows an old, leaked cert for some domain). I guess users would need to ship the applications with a reasonably up-to-date revocation lists?

Proactive measurements: (not a risk per se, but will make it harder to introduce new bugs)
- [ ] Run tests with sanitizers in CI (see #2106).
- [ ] Run some linters in CI.
- [ ] Refactorings and simplifying Graphene code, especially the old parts and old, bug-prone APIs.

**Comment by 0400H** [comment]
@dimakuv 

[protected files](https://github.com/oscarlab/graphene/pull/1325) is claimed that integrity of data is protected. How is it achieved? Is it a check of message digest?

**Comment by mkow** [comment]
Message? There are no messages there ;)
Anyways, we use authenticated encryption (precisely: AES-GCM), which provides both confidentiality and integrity.

**Comment by 0400H** [comment]
> Message? There are no messages there ;)
> Anyways, we use authenticated encryption (precisely: AES-GCM), which provides both confidentiality and integrity.

@mkow 

Thank you for your reply.  :-)
A [message digest](https://www.techopedia.com/definition/4024/message-digest) is a cryptographic hash function containing a string of digits created by a one-way hashing formula. :-)

I found it. :-)
[pf_crypt.c:pf_init](https://github.com/oscarlab/graphene/blob/1102fee8f5f5ea85316b1d20a736500b0f1bc04d/Pal/src/host/Linux-SGX/tools/pf_crypt/pf_crypt.c#L95) --> [pf_util.c:mbedtls_aes_gcm_encrypt](https://github.com/oscarlab/graphene/blob/1102fee8f5f5ea85316b1d20a736500b0f1bc04d/Pal/src/host/Linux-SGX/tools/common/pf_util.c#L110)

**Comment by mkow** [comment]
Yeah, I know what's a hash/digest :) I was just confused by "message", which makes no sense in this context.

> I found it. :-)

Yup, that's the place :)

Btw. If you want to continue this discussion please move it to our mailing list, or if you find a bug then create a separate issue, otherwise it's a bit off-topic here (this is a meta-issue to keep track of the subtasks).
